gcc/rust/lint/rust-lint-marklive.cc

```diff
@@ -91,31 +91,39 @@ MarkLive::go (HIR::Crate &crate)
 void
 MarkLive::visit (HIR::PathInExpression &expr)
 {
-  NodeId ast_node_id = expr.get_mappings ().get_nodeid ();
-  NodeId ref_node_id = UNKNOWN_NODEID;
-  if (resolver->lookup_resolved_name (ast_node_id, &ref_node_id))
-    {
-      Resolver::Definition def;
-      if (!resolver->lookup_definition (ref_node_id, &def))
-	{
-	  rust_error_at (expr.get_locus (),
-			 "unknown reference for resolved name");
-	  return;
-	}
-      ref_node_id = def.parent;
-      HirId ref;
-      if (!mappings->lookup_node_to_hir (expr.get_mappings ().get_crate_num (),
-					 ref_node_id, &ref))
-	{
-	  rust_error_at (expr.get_locus (), "reverse lookup failure");
-	  return;
-	}
-      if (scannedSymbols.find (ref) == scannedSymbols.end ())
-	{
-	  worklist.push_back (ref);
-	}
-      liveSymbols.emplace (ref);
-    }
+  expr.iterate_path_segments ([&] (HIR::PathExprSegment &seg) -> bool {
```

Comment:
I think you got it right here. Baiscaly PathExpressions are:

A::B::C, but they might also contain generic arguments per segment. So when we do path resolution we iterate each segment for type resolution. When it comes to compilation to GENERIC this is simplified because:

Segment {
  mappings,
  path
}

PathExpr {
  list<Segment>
  mappings
}

So this is a simplified version of the path expression, but once type resolution runs it should be possible to simply call all this code once based on the hir mappings for the whole PathInExpression. 

I don't think you need to change anything here, lets see how this works for now.

Reply:
https://github.com/Rust-GCC/gccrs/blob/bab6f6bcb299ac42096783cda32b8f11da4288ca/gcc/rust/backend/rust-compile-resolve-path.cc#L31